src/source_distribution.rs

```diff
@@ -139,7 +143,38 @@ fn add_crate_to_source_distribution(
     Ok(())
 }
 
-/// Creates aif source distribution, packing the root crate and all local dependencies
+/// Finds all path dependencies of the crate.
+fn find_path_deps(cargo_metadata: &Metadata) -> Result<HashMap<&String, &PathBuf>> {
+    let root = cargo_metadata
+        .root_package()
+        .context("Expected the dependency graph to have a root package")?;
+    // scan the dependency graph for path dependencies
+    let mut path_deps = HashMap::new();
+    let mut stack: Vec<&cargo_metadata::Package> = vec![root];
+    while let Some(top) = stack.pop() {
+        for dependency in &top.dependencies {
+            if let Some(path) = &dependency.path {
+                path_deps.insert(&dependency.name, path);
+                // we search for the respective package by `manifest_path`, there seems
+                // to be no way to query the dependency graph given `dependency`
```

Comment:
The cargo metadata output for the root package contains everything you need to know, no need to run it multiple times (see resolve.nodes)

Reply:
I am aware of `resolve.nodes`. Hower, the [`Dependency`](https://docs.rs/cargo_metadata/0.12.3/cargo_metadata/struct.Dependency.html) obtained by iterating over `top.dependencies` does not contain a [`PackageId`](https://docs.rs/cargo_metadata/0.12.3/cargo_metadata/struct.PackageId.html) field which would be necessary to locate the respective dependency in `cargo_metadata.packages` or in `resolve.nodes`. Also, the dependency graph is transitivity reduced and, as I argued here https://github.com/PyO3/maturin/pull/453#issuecomment-788652178, just iterating over its nodes does include unnecessary crates into the source distribution. Hence, it seems to me, one needs to actually traverse the graph to find all the actual (transitive) path dependencies of the crate.

Now, traversing `resolve.nodes` has an output-sensitive complexity of *O(nm)* where *n* is the amount of (transient) path dependencies and *m* is the number of nodes of the graph. As `resolve.nodes` is a `Vec` we have to scan it linearly given the package ids obtained via a node's [`Node.dependencies`](https://docs.rs/cargo_metadata/0.12.3/cargo_metadata/struct.Node.html#structfield.dependencies). This is the same complexity as my proposed change. What would help, though, is to first build a data structure supporting more efficient queries for dependencies than linearly scanning. Maybe, I miss something here but I cannot see how `resolve.nodes` helps—it represents the dependency graph but not in a way that enables efficient queries for (transient) path dependencies.

In addition, working with [`Dependency`](https://docs.rs/cargo_metadata/0.12.3/cargo_metadata/struct.Dependency.html) has the benefit that it is clearly documented how to identify path dependencies. In contrast, the `repr` field of [`PackageId`](https://docs.rs/cargo_metadata/0.12.3/cargo_metadata/struct.PackageId.html) is explicitly considered an implementation detail.